In [1]:
import quandl

In [135]:
quandl.ApiConfig.api_key = '{yourkeyhere}'

In [136]:
data = quandl.get("TFGRAIN/SOYBEANS", authtoken="{yourkeyhere}")

In [137]:
data.tail()

Cash Price  Basis  Fall Price  Fall Basis
Date                                                 
2019-01-08        8.56  -0.63        9.13       -0.45
2019-01-09        8.61  -0.63        9.18       -0.45
2019-01-10        8.45  -0.62        9.03       -0.45
2019-01-11        8.48  -0.62        9.07       -0.45
2019-01-14        8.42  -0.62        9.02       -0.45

In [14]:
len(data)

4535

In [138]:
%load_ext autoreload
%autoreload 2
%matplotlib inline

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [139]:
from fastai.imports import *
from fastai.structured import *
from pandas_summary import DataFrameSummary
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier
from IPython.display import display
from sklearn import metrics

In [140]:
data.rename_axis('Date')

Cash Price  Basis  Fall Price  Fall Basis
Date                                                 
2000-12-01        4.90  -0.18        5.02       -0.26
2000-12-04        4.87  -0.18        4.99       -0.26
2000-12-05        4.83  -0.18        4.98       -0.26
2000-12-06        4.89  -0.18        5.02       -0.26
2000-12-07        4.85  -0.18        4.98       -0.26
2000-12-08        4.94  -0.18        5.03       -0.26
2000-12-11        4.93  -0.18        5.00       -0.26
2000-12-12        4.89  -0.19        4.96       -0.26
2000-12-13        4.88  -0.19        4.97       -0.26
2000-12-14        4.86  -0.19        4.97       -0.26
2000-12-15        4.89  -0.19        4.99       -0.26
2000-12-18        4.91  -0.19        5.02       -0.26
2000-12-19        4.93  -0.19        5.03       -0.26
2000-12-20        4.91  -0.17        5.00       -0.26
2000-12-21        4.95  -0.15        5.01       -0.26
2000-12-22        4.96  -0.15        4.98       -0.26
2000-12-26        4.91  -0.15        4.98       -0.26
2000-12-27        4.86  -0.13        4.93       -0.26
2000-12-28        4.85  -0.13        4.89       -0.26
2000-12-29        4.87  -0.13        4.91       -0.26
2001-01-02        4.82  -0.13        4.88       -0.26
2001-01-03        4.85  -0.13        4.89       -0.26
2001-01-04        4.80  -0.21        4.84       -0.26
2001-01-05        4.77  -0.21        4.81       -0.26
2001-01-08        4.78  -0.21        4.82       -0.26
2001-01-09        4.82  -0.21        4.85       -0.26
2001-01-10        4.77  -0.21        4.81       -0.26
2001-01-11        4.64  -0.21        4.71       -0.26
2001-01-12        4.62  -0.21        4.67       -0.26
2001-01-16        4.59  -0.19        4.60       -0.26
...                ...    ...         ...         ...
2018-11-30        8.40  -0.55        8.94       -0.45
2018-12-03        8.51  -0.55        9.05       -0.45
2018-12-04        8.57  -0.55        9.11       -0.45
2018-12-05        8.59  -0.55        9.12       -0.45
2018-12-06        8.55  -0.55        9.10       -0.45
2018-12-07        8.62  -0.55        9.16       -0.45
2018-12-10        8.55  -0.55        9.12       -0.45
2018-12-11        8.60  -0.55        9.17       -0.45
2018-12-12        8.65  -0.55        9.21       -0.45
2018-12-13        8.52  -0.55        9.11       -0.45
2018-12-14        8.46  -0.55        9.06       -0.45
2018-12-17        8.50  -0.55        9.09       -0.45
2018-12-18        8.53  -0.55        9.11       -0.45
2018-12-19        8.45  -0.55        9.04       -0.45
2018-12-20        8.39  -0.55        8.98       -0.45
2018-12-21        8.30  -0.55        8.92       -0.45
2018-12-24        8.29  -0.55        8.92       -0.45
2018-12-26        8.15  -0.55        8.81       -0.45
2018-12-27        8.14  -0.55        8.80       -0.45
2018-12-28        8.28  -0.55        8.91       -0.45
2018-12-31        8.32  -0.63        8.90       -0.45
2019-01-02        8.44  -0.63        8.99       -0.45
2019-01-03        8.50  -0.63        9.04       -0.45
2019-01-04        8.59  -0.63        9.12       -0.45
2019-01-07        8.61  -0.63        9.16       -0.45
2019-01-08        8.56  -0.63        9.13       -0.45
2019-01-09        8.61  -0.63        9.18       -0.45
2019-01-10        8.45  -0.62        9.03       -0.45
2019-01-11        8.48  -0.62        9.07       -0.45
2019-01-14        8.42  -0.62        9.02       -0.45

[4535 rows x 4 columns]

In [157]:
data['Date'] = data.index
add_datepart(data, 'Date')
data.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 4535 entries, 2000-12-01 to 2019-01-14
Data columns (total 18 columns):
id                  4535 non-null int64
Cash Price          4535 non-null float64
Basis               4535 non-null float64
Fall Price          4516 non-null float64
Fall Basis          4516 non-null float64
Year                4535 non-null int64
Month               4535 non-null int64
Week                4535 non-null int64
Day                 4535 non-null int64
Dayofweek           4535 non-null int64
Dayofyear           4535 non-null int64
Is_month_end        4535 non-null bool
Is_month_start      4535 non-null bool
Is_quarter_end      4535 non-null bool
Is_quarter_start    4535 non-null bool
Is_year_end         4535 non-null bool
Is_year_start       4535 non-null bool
Elapsed             4535 non-null int64
dtypes: bool(6), float64(4), int64(8)
memory usage: 487.2 KB


In [159]:
##not working, see inelegant below
dfdata = data.reset_index()
dfdata.columns[0] = 'Id'
dfdata['Id'] = dfdata.index
dfdata.info()
#dfdata = data
#dfdata.insert(0,'id', range(0,len(dfdata)))
#dfdata.set_index('id')

TypeError: Index does not support mutable operations

In [161]:
col_list = ['Cash Price', 'Basis', 'Year', 'Month', 'Week', 'Day', 'Dayofweek', 'Dayofyear']
dfdata = dfdata[col_list]

In [162]:
npdf = dfdata.values

In [163]:
npdf

array([[   4.9 ,   -0.18, 2000.  , ...,    1.  ,    4.  ,  336.  ],
       [   4.87,   -0.18, 2000.  , ...,    4.  ,    0.  ,  339.  ],
       [   4.83,   -0.18, 2000.  , ...,    5.  ,    1.  ,  340.  ],
       ...,
       [   8.45,   -0.62, 2019.  , ...,   10.  ,    3.  ,   10.  ],
       [   8.48,   -0.62, 2019.  , ...,   11.  ,    4.  ,   11.  ],
       [   8.42,   -0.62, 2019.  , ...,   14.  ,    0.  ,   14.  ]])

In [164]:
fnpdf = pd.DataFrame.from_records(npdf)

In [165]:
fnpdf

0     1       2     3     4     5    6      7
0     4.90 -0.18  2000.0  12.0  48.0   1.0  4.0  336.0
1     4.87 -0.18  2000.0  12.0  49.0   4.0  0.0  339.0
2     4.83 -0.18  2000.0  12.0  49.0   5.0  1.0  340.0
3     4.89 -0.18  2000.0  12.0  49.0   6.0  2.0  341.0
4     4.85 -0.18  2000.0  12.0  49.0   7.0  3.0  342.0
5     4.94 -0.18  2000.0  12.0  49.0   8.0  4.0  343.0
6     4.93 -0.18  2000.0  12.0  50.0  11.0  0.0  346.0
7     4.89 -0.19  2000.0  12.0  50.0  12.0  1.0  347.0
8     4.88 -0.19  2000.0  12.0  50.0  13.0  2.0  348.0
9     4.86 -0.19  2000.0  12.0  50.0  14.0  3.0  349.0
10    4.89 -0.19  2000.0  12.0  50.0  15.0  4.0  350.0
11    4.91 -0.19  2000.0  12.0  51.0  18.0  0.0  353.0
12    4.93 -0.19  2000.0  12.0  51.0  19.0  1.0  354.0
13    4.91 -0.17  2000.0  12.0  51.0  20.0  2.0  355.0
14    4.95 -0.15  2000.0  12.0  51.0  21.0  3.0  356.0
15    4.96 -0.15  2000.0  12.0  51.0  22.0  4.0  357.0
16    4.91 -0.15  2000.0  12.0  52.0  26.0  1.0  361.0
17    4.86 -0.13  2000.0  12.0  52.0  27.0  2.0  362.0
18    4.85 -0.13  2000.0  12.0  52.0  28.0  3.0  363.0
19    4.87 -0.13  2000.0  12.0  52.0  29.0  4.0  364.0
20    4.82 -0.13  2001.0   1.0   1.0   2.0  1.0    2.0
21    4.85 -0.13  2001.0   1.0   1.0   3.0  2.0    3.0
22    4.80 -0.21  2001.0   1.0   1.0   4.0  3.0    4.0
23    4.77 -0.21  2001.0   1.0   1.0   5.0  4.0    5.0
24    4.78 -0.21  2001.0   1.0   2.0   8.0  0.0    8.0
25    4.82 -0.21  2001.0   1.0   2.0   9.0  1.0    9.0
26    4.77 -0.21  2001.0   1.0   2.0  10.0  2.0   10.0
27    4.64 -0.21  2001.0   1.0   2.0  11.0  3.0   11.0
28    4.62 -0.21  2001.0   1.0   2.0  12.0  4.0   12.0
29    4.59 -0.19  2001.0   1.0   3.0  16.0  1.0   16.0
...    ...   ...     ...   ...   ...   ...  ...    ...
4505  8.40 -0.55  2018.0  11.0  48.0  30.0  4.0  334.0
4506  8.51 -0.55  2018.0  12.0  49.0   3.0  0.0  337.0
4507  8.57 -0.55  2018.0  12.0  49.0   4.0  1.0  338.0
4508  8.59 -0.55  2018.0  12.0  49.0   5.0  2.0  339.0
4509  8.55 -0.55  2018.0  12.0  49.0   6.0  3.0  340.0
4510  8.62 -0.55  2018.0  12.0  49.0   7.0  4.0  341.0
4511  8.55 -0.55  2018.0  12.0  50.0  10.0  0.0  344.0
4512  8.60 -0.55  2018.0  12.0  50.0  11.0  1.0  345.0
4513  8.65 -0.55  2018.0  12.0  50.0  12.0  2.0  346.0
4514  8.52 -0.55  2018.0  12.0  50.0  13.0  3.0  347.0
4515  8.46 -0.55  2018.0  12.0  50.0  14.0  4.0  348.0
4516  8.50 -0.55  2018.0  12.0  51.0  17.0  0.0  351.0
4517  8.53 -0.55  2018.0  12.0  51.0  18.0  1.0  352.0
4518  8.45 -0.55  2018.0  12.0  51.0  19.0  2.0  353.0
4519  8.39 -0.55  2018.0  12.0  51.0  20.0  3.0  354.0
4520  8.30 -0.55  2018.0  12.0  51.0  21.0  4.0  355.0
4521  8.29 -0.55  2018.0  12.0  52.0  24.0  0.0  358.0
4522  8.15 -0.55  2018.0  12.0  52.0  26.0  2.0  360.0
4523  8.14 -0.55  2018.0  12.0  52.0  27.0  3.0  361.0
4524  8.28 -0.55  2018.0  12.0  52.0  28.0  4.0  362.0
4525  8.32 -0.63  2018.0  12.0   1.0  31.0  0.0  365.0
4526  8.44 -0.63  2019.0   1.0   1.0   2.0  2.0    2.0
4527  8.50 -0.63  2019.0   1.0   1.0   3.0  3.0    3.0
4528  8.59 -0.63  2019.0   1.0   1.0   4.0  4.0    4.0
4529  8.61 -0.63  2019.0   1.0   2.0   7.0  0.0    7.0
4530  8.56 -0.63  2019.0   1.0   2.0   8.0  1.0    8.0
4531  8.61 -0.63  2019.0   1.0   2.0   9.0  2.0    9.0
4532  8.45 -0.62  2019.0   1.0   2.0  10.0  3.0   10.0
4533  8.48 -0.62  2019.0   1.0   2.0  11.0  4.0   11.0
4534  8.42 -0.62  2019.0   1.0   3.0  14.0  0.0   14.0

[4535 rows x 8 columns]

In [71]:
#fnpdf.drop([10], inplace=True, axis=1)

In [166]:
fnpdf.head()

0     1       2     3     4    5    6      7
0  4.90 -0.18  2000.0  12.0  48.0  1.0  4.0  336.0
1  4.87 -0.18  2000.0  12.0  49.0  4.0  0.0  339.0
2  4.83 -0.18  2000.0  12.0  49.0  5.0  1.0  340.0
3  4.89 -0.18  2000.0  12.0  49.0  6.0  2.0  341.0
4  4.85 -0.18  2000.0  12.0  49.0  7.0  3.0  342.0

In [167]:
fnpdf.columns = col_list
fnpdf

Cash Price  Basis    Year  Month  Week   Day  Dayofweek  Dayofyear
0           4.90  -0.18  2000.0   12.0  48.0   1.0        4.0      336.0
1           4.87  -0.18  2000.0   12.0  49.0   4.0        0.0      339.0
2           4.83  -0.18  2000.0   12.0  49.0   5.0        1.0      340.0
3           4.89  -0.18  2000.0   12.0  49.0   6.0        2.0      341.0
4           4.85  -0.18  2000.0   12.0  49.0   7.0        3.0      342.0
5           4.94  -0.18  2000.0   12.0  49.0   8.0        4.0      343.0
6           4.93  -0.18  2000.0   12.0  50.0  11.0        0.0      346.0
7           4.89  -0.19  2000.0   12.0  50.0  12.0        1.0      347.0
8           4.88  -0.19  2000.0   12.0  50.0  13.0        2.0      348.0
9           4.86  -0.19  2000.0   12.0  50.0  14.0        3.0      349.0
10          4.89  -0.19  2000.0   12.0  50.0  15.0        4.0      350.0
11          4.91  -0.19  2000.0   12.0  51.0  18.0        0.0      353.0
12          4.93  -0.19  2000.0   12.0  51.0  19.0        1.0      354.0
13          4.91  -0.17  2000.0   12.0  51.0  20.0        2.0      355.0
14          4.95  -0.15  2000.0   12.0  51.0  21.0        3.0      356.0
15          4.96  -0.15  2000.0   12.0  51.0  22.0        4.0      357.0
16          4.91  -0.15  2000.0   12.0  52.0  26.0        1.0      361.0
17          4.86  -0.13  2000.0   12.0  52.0  27.0        2.0      362.0
18          4.85  -0.13  2000.0   12.0  52.0  28.0        3.0      363.0
19          4.87  -0.13  2000.0   12.0  52.0  29.0        4.0      364.0
20          4.82  -0.13  2001.0    1.0   1.0   2.0        1.0        2.0
21          4.85  -0.13  2001.0    1.0   1.0   3.0        2.0        3.0
22          4.80  -0.21  2001.0    1.0   1.0   4.0        3.0        4.0
23          4.77  -0.21  2001.0    1.0   1.0   5.0        4.0        5.0
24          4.78  -0.21  2001.0    1.0   2.0   8.0        0.0        8.0
25          4.82  -0.21  2001.0    1.0   2.0   9.0        1.0        9.0
26          4.77  -0.21  2001.0    1.0   2.0  10.0        2.0       10.0
27          4.64  -0.21  2001.0    1.0   2.0  11.0        3.0       11.0
28          4.62  -0.21  2001.0    1.0   2.0  12.0        4.0       12.0
29          4.59  -0.19  2001.0    1.0   3.0  16.0        1.0       16.0
...          ...    ...     ...    ...   ...   ...        ...        ...
4505        8.40  -0.55  2018.0   11.0  48.0  30.0        4.0      334.0
4506        8.51  -0.55  2018.0   12.0  49.0   3.0        0.0      337.0
4507        8.57  -0.55  2018.0   12.0  49.0   4.0        1.0      338.0
4508        8.59  -0.55  2018.0   12.0  49.0   5.0        2.0      339.0
4509        8.55  -0.55  2018.0   12.0  49.0   6.0        3.0      340.0
4510        8.62  -0.55  2018.0   12.0  49.0   7.0        4.0      341.0
4511        8.55  -0.55  2018.0   12.0  50.0  10.0        0.0      344.0
4512        8.60  -0.55  2018.0   12.0  50.0  11.0        1.0      345.0
4513        8.65  -0.55  2018.0   12.0  50.0  12.0        2.0      346.0
4514        8.52  -0.55  2018.0   12.0  50.0  13.0        3.0      347.0
4515        8.46  -0.55  2018.0   12.0  50.0  14.0        4.0      348.0
4516        8.50  -0.55  2018.0   12.0  51.0  17.0        0.0      351.0
4517        8.53  -0.55  2018.0   12.0  51.0  18.0        1.0      352.0
4518        8.45  -0.55  2018.0   12.0  51.0  19.0        2.0      353.0
4519        8.39  -0.55  2018.0   12.0  51.0  20.0        3.0      354.0
4520        8.30  -0.55  2018.0   12.0  51.0  21.0        4.0      355.0
4521        8.29  -0.55  2018.0   12.0  52.0  24.0        0.0      358.0
4522        8.15  -0.55  2018.0   12.0  52.0  26.0        2.0      360.0
4523        8.14  -0.55  2018.0   12.0  52.0  27.0        3.0      361.0
4524        8.28  -0.55  2018.0   12.0  52.0  28.0        4.0      362.0
4525        8.32  -0.63  2018.0   12.0   1.0  31.0        0.0      365.0
4526        8.44  -0.63  2019.0    1.0   1.0   2.0        2.0        2.0
4527        8.50  -0.63  2019.0    1.0   1.0   3.0        3.0   

In [168]:
train_cats(fnpdf)

In [170]:
df, y, nas = proc_df(fnpdf, 'Cash Price')

In [171]:
df

Basis    Year  Month  Week   Day  Dayofweek  Dayofyear
0     -0.18  2000.0   12.0  48.0   1.0        4.0      336.0
1     -0.18  2000.0   12.0  49.0   4.0        0.0      339.0
2     -0.18  2000.0   12.0  49.0   5.0        1.0      340.0
3     -0.18  2000.0   12.0  49.0   6.0        2.0      341.0
4     -0.18  2000.0   12.0  49.0   7.0        3.0      342.0
5     -0.18  2000.0   12.0  49.0   8.0        4.0      343.0
6     -0.18  2000.0   12.0  50.0  11.0        0.0      346.0
7     -0.19  2000.0   12.0  50.0  12.0        1.0      347.0
8     -0.19  2000.0   12.0  50.0  13.0        2.0      348.0
9     -0.19  2000.0   12.0  50.0  14.0        3.0      349.0
10    -0.19  2000.0   12.0  50.0  15.0        4.0      350.0
11    -0.19  2000.0   12.0  51.0  18.0        0.0      353.0
12    -0.19  2000.0   12.0  51.0  19.0        1.0      354.0
13    -0.17  2000.0   12.0  51.0  20.0        2.0      355.0
14    -0.15  2000.0   12.0  51.0  21.0        3.0      356.0
15    -0.15  2000.0   12.0  51.0  22.0        4.0      357.0
16    -0.15  2000.0   12.0  52.0  26.0        1.0      361.0
17    -0.13  2000.0   12.0  52.0  27.0        2.0      362.0
18    -0.13  2000.0   12.0  52.0  28.0        3.0      363.0
19    -0.13  2000.0   12.0  52.0  29.0        4.0      364.0
20    -0.13  2001.0    1.0   1.0   2.0        1.0        2.0
21    -0.13  2001.0    1.0   1.0   3.0        2.0        3.0
22    -0.21  2001.0    1.0   1.0   4.0        3.0        4.0
23    -0.21  2001.0    1.0   1.0   5.0        4.0        5.0
24    -0.21  2001.0    1.0   2.0   8.0        0.0        8.0
25    -0.21  2001.0    1.0   2.0   9.0        1.0        9.0
26    -0.21  2001.0    1.0   2.0  10.0        2.0       10.0
27    -0.21  2001.0    1.0   2.0  11.0        3.0       11.0
28    -0.21  2001.0    1.0   2.0  12.0        4.0       12.0
29    -0.19  2001.0    1.0   3.0  16.0        1.0       16.0
...     ...     ...    ...   ...   ...        ...        ...
4505  -0.55  2018.0   11.0  48.0  30.0        4.0      334.0
4506  -0.55  2018.0   12.0  49.0   3.0        0.0      337.0
4507  -0.55  2018.0   12.0  49.0   4.0        1.0      338.0
4508  -0.55  2018.0   12.0  49.0   5.0        2.0      339.0
4509  -0.55  2018.0   12.0  49.0   6.0        3.0      340.0
4510  -0.55  2018.0   12.0  49.0   7.0        4.0      341.0
4511  -0.55  2018.0   12.0  50.0  10.0        0.0      344.0
4512  -0.55  2018.0   12.0  50.0  11.0        1.0      345.0
4513  -0.55  2018.0   12.0  50.0  12.0        2.0      346.0
4514  -0.55  2018.0   12.0  50.0  13.0        3.0      347.0
4515  -0.55  2018.0   12.0  50.0  14.0        4.0      348.0
4516  -0.55  2018.0   12.0  51.0  17.0        0.0      351.0
4517  -0.55  2018.0   12.0  51.0  18.0        1.0      352.0
4518  -0.55  2018.0   12.0  51.0  19.0        2.0      353.0
4519  -0.55  2018.0   12.0  51.0  20.0        3.0      354.0
4520  -0.55  2018.0   12.0  51.0  21.0        4.0      355.0
4521  -0.55  2018.0   12.0  52.0  24.0        0.0      358.0
4522  -0.55  2018.0   12.0  52.0  26.0        2.0      360.0
4523  -0.55  2018.0   12.0  52.0  27.0        3.0      361.0
4524  -0.55  2018.0   12.0  52.0  28.0        4.0      362.0
4525  -0.63  2018.0   12.0   1.0  31.0        0.0      365.0
4526  -0.63  2019.0    1.0   1.0   2.0        2.0        2.0
4527  -0.63  2019.0    1.0   1.0   3.0        3.0        3.0
4528  -0.63  2019.0    1.0   1.0   4.0        4.0        4.0
4529  -0.63  2019.0    1.0   2.0   7.0        0.0        7.0
4530  -0.63  2019.0    1.0   2.0   8.0        1.0        8.0
4531  -0.63  2019.0    1.0   2.0   9.0        2.0        9.0
4532  -0.62  2019.0    1.0   2.0  10.0        3.0       10.0
4533  -0.62  2019.0    1.0   2.0  11.0        4.0       11.0
4534  -0.62  2019.0    1.0   3.0  14.0        0.0       14.0

[4535 rows x 7 columns]

In [172]:
m = RandomForestRegressor(n_jobs=-1)
m.fit(df, y)
m.score(df,y)

/Users/matthew.arthur/anaconda3/envs/fastai-cpu/lib/python3.6/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


0.9991621993655437

In [173]:
df.head(1)

Basis    Year  Month  Week  Day  Dayofweek  Dayofyear
0  -0.18  2000.0   12.0  48.0  1.0        4.0      336.0

In [174]:
df1 = pd.DataFrame({'Basis':[-.75],
                    'Year':[2019],
                    'Month':[1],
                    'Week':[3],
                    'Day':[16],
                    'Dayofweek':[4],
                    'Dayofyear':[16],})
df1

Basis  Year  Month  Week  Day  Dayofweek  Dayofyear
0  -0.75  2019      1     3   16          4         16

In [175]:
train_cats(df1)

In [176]:
m.predict(df1)

array([8.465])

Today's soybean cash price is 8.21 at Pilot Grove versus predicted value of 8.465.